subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1278 @@
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+
+	LOG_DBG("");
+	desc->if0_ncm.iMACAddress = 0;
+	sys_dlist_remove(&data->mac_desc_data->node);
+}   /* usbd_cdc_ncm_shutdown */
+
+
+
+static void *usbd_cdc_ncm_get_desc(struct usbd_class_data *const c_data,
+		const enum usbd_speed speed)
+/**
+ * Get function descriptor based on speed parameter
+ */
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *const data = dev->data;
+
+	if (speed == USBD_SPEED_HS) {
+		return data->hs_desc;
+	}
+
+	return data->fs_desc;
+}   /* usbd_cdc_ncm_get_desc */
+
+
+
+static int cdc_ncm_send(const struct device *dev, struct net_pkt *const pkt)
+/**
+ * Send a network packet to the host
+ */
+{
+	struct cdc_ncm_eth_data *const data = dev->data;
+	struct usbd_class_data *c_data = data->c_data;
+	size_t pkt_len = net_pkt_get_len(pkt);
+	struct net_buf *buf;
+	union xmit_ntb_t *ntb;
+
+	LOG_DBG("len: %d", pkt_len);
+
+	if (pkt_len > sizeof(ntb->data)) {
+		LOG_WRN("Trying to send too large packet, drop");
+		return -ENOMEM;
+	}
+
+	if (!atomic_test_bit(&data->state, CDC_NCM_CLASS_ENABLED)
+	    ||  !atomic_test_bit(&data->state, CDC_NCM_IFACE_UP)) {
+		LOG_INF("Configuration is not enabled or interface not ready %ld", data->state);
+		return -EACCES;
+	}
+
+	buf = cdc_ncm_buf_alloc(cdc_ncm_get_bulk_in(c_data));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return -ENOMEM;
+	}
+
+	/*
+	 * create (simple) NTB
+	 */
+	ntb = (union xmit_ntb_t *)buf->data;
+
+	ntb->nth.dwSignature   = sys_cpu_to_le32(NTH16_SIGNATURE);
+	ntb->nth.wHeaderLength = sys_cpu_to_le16(sizeof(struct nth16_t));
+	ntb->nth.wSequence     = sys_cpu_to_le16(++data->tx_sequence);
+	ntb->nth.wNdpIndex     = sys_cpu_to_le16(sizeof(struct nth16_t));
+
+	ntb->ndp.dwSignature   = sys_cpu_to_le32(NDP16_SIGNATURE_NCM0);
+	ntb->ndp.wLength       = sys_cpu_to_le16(sizeof(struct ndp16_t)
+		+ (CFG_CDC_NCM_XMT_MAX_DATAGRAMS_PER_NTB + 1)*sizeof(struct ndp16_datagram_t));
+	ntb->ndp.wNextNdpIndex = 0;
+
+	ntb->ndp_datagram[0].wDatagramIndex  =
+			sys_cpu_to_le16(sys_le16_to_cpu(ntb->nth.wHeaderLength)
+					+ sys_le16_to_cpu(ntb->ndp.wLength));
+	ntb->ndp_datagram[0].wDatagramLength = sys_cpu_to_le16(pkt_len);
+	ntb->ndp_datagram[1].wDatagramIndex  = 0;
+	ntb->ndp_datagram[1].wDatagramLength = 0;
+
+	ntb->nth.wBlockLength =
+			sys_cpu_to_le16(sys_le16_to_cpu(ntb->ndp_datagram[0].wDatagramIndex)
+					+ pkt_len);
+
+	if (net_pkt_read(pkt,
+		ntb->data + sys_le16_to_cpu(ntb->ndp_datagram[0].wDatagramIndex), pkt_len)) {
+		LOG_ERR("Failed copy net_pkt");
+		net_buf_unref(buf);
+		return -ENOBUFS;
+	}
+	/* post: now the complete NTB is written. */
+
+	/* adjust length of net_buf accordingly */
+	net_buf_add(buf, sys_le16_to_cpu(ntb->nth.wBlockLength));
+
+	if (sys_le16_to_cpu(ntb->nth.wBlockLength) % cdc_ncm_get_bulk_in_mps(c_data) == 0) {
+		udc_ep_buf_set_zlp(buf);
+	}
+
+	usbd_ep_enqueue(c_data, buf);
+	k_sem_take(&data->sync_sem, K_FOREVER);
```

Comment:
Could you hint me why this is required? It seems released here: https://github.com/zephyrproject-rtos/zephyr/pull/72310/files#diff-d6af60862ee4693cfe35b5df0806df2fdfff1921acc16a65b23ee407548e7212R598

If it is for managing the concurrency that `usbd_ep_enqueue()` might generate, should it be placed one line before to handle a context switch between `usbd_ep_enqueue()` and `k_sem_take()`?

I think I simply misunderstood something.

Reply:
Again: taken directly from ECM.  It blocks the thread in line 930 until the semaphore is given in the completion event handler usbd_cdc_ncm_request().

As said, the driver is currently very simple and not performance trimmed.

Reply:
Thank you! This suggests it was already reviewed/validated earlier.